File: examples/complex/artifact-pipeline/python/_cott_impl/curriculum/artifact_pipeline/plan_pipeline.py
from cott_runtime import Err, Ok, Result
from curriculum.artifact_pipeline import topologically_order_steps
from curriculum.artifact_pipeline_types import ArtifactPipelineError, ArtifactPlan, Pipeline


def plan_pipeline(pipeline: Pipeline) -> Result[ArtifactPlan, ArtifactPipelineError]:
    match topologically_order_steps(pipeline.steps):
        case Ok(value=ordered_steps):
            return Ok(value=ArtifactPlan(ordered_steps=ordered_steps))
        case Err(error=error):
            return Err(error=error)

File: examples/kotlin/complex/artifact-pipeline/kotlin/cott_impl/curriculum/artifact_pipeline/plan_pipeline.kt
package cott_impl.curriculum.artifact_pipeline

internal fun plan_pipeline(pipeline: curriculum.artifact_pipeline.Pipeline): cott_runtime.CottResult<curriculum.artifact_pipeline.ArtifactPlan, curriculum.artifact_pipeline.ArtifactPipelineError> {
    return when (val ordered = curriculum.artifact_pipeline.topologically_order_steps(pipeline.steps)) {
        is cott_runtime.Ok -> cott_runtime.Ok(curriculum.artifact_pipeline.ArtifactPlan(ordered.value))
        is cott_runtime.Err -> ordered
    }
}
